Aplib.Core/Intent/Actions/Action.cs

```diff
@@ -21,24 +26,42 @@ public class Action
         /// Initializes a new empty instance of the <see cref="Action{TQuery}"/> class.
         /// </summary>
         /// <remarks>Only meant for internal use</remarks>
-        protected internal Action()
+        /// <param name="name">The name of this Action, used to quickly display this goal in several contexts.</param>
+        /// <param name="description">
+        /// The description of this Action, used to explain this goal in several contexts.
+        /// </param>
+        protected internal Action(string name, string? description = null)
```

Comment:
I am not a fan of a required name. If the name is for debugging purposes, I would say it must be possible to not name the action.

If one wants to change the name of the action, it can do so via the metadata:

```cs
Action action = new(() => {Console.WriteLine("It's gonna be Legen... wait for it... DARY");});
action.Metadata.Name = "Barney's Action";
action.Metadata.Description = "FRIENDSHIP OVER";
```

Furthermore, if we want to add something to the metadata, we would need to alter all constructors. Therefore, I would propose that we don't ask for the name and description, but for the user to supply the metadata object themselves:

```cs
Action action = new
(
    () => Console.WriteLine("It's gonna be Legen... wait for it... DARY"), 
    new Metadata(name: "Barney's Action", description: "FRIENDSHIP OVER")
);
```

Reply:
Yes I completely agree, that sounds way better. I will change it.

Reply:
About the required name though. I think there are thee ways to make the name not required.

1. Make the Metadata field an optional.
2. Make the Name field in Metadata optional.
3. Add a GUID to the metadata object.

I don't like really like the first option but it is possible.

I think the second option is not great, because then an object may have a Metadata attached that is non-null, but only containing null data.

I kind of like the last option, because it will also later on give us an easy way to distinguish components with certainty. And GUIDs are built into C# so it is easy.

Reply:
Great options! I think the second one is the way to go. 

You could create a new Metadata object everytime you instance the class using default properties. You could also have a constructor which can have an optional parameter `metadata` which if it is null, will be instanced to an empty Metadata class. This way, we always have a way to provide metadata in the constructor, but also outside the constructor.

Reply:
We could add a GUID to the metadata object. That is actually a great idea, and can be used as a sort of 'default' name. However, do make the GUID read-only.